reportportal_client/core/worker.py

```diff
@@ -0,0 +1,169 @@
+"""This module contains worker that makes non-blocking HTTP requests.
+
+Copyright (c) 2018 http://reportportal.io .
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+from enum import auto, Enum, unique
+import logging
+from threading import currentThread, Thread
+from queue import Empty
+
+logger = logging.getLogger(__name__)
+logger.addHandler(logging.NullHandler())
+
+
+@unique
+class ControlCommand(Enum):
+    """This class stores worker control commands."""
+
+    CLEAR_QUEUE = auto()
+    NOP = auto()
+    REPORT_STATUS = auto()
+    STOP = auto()
+    STOP_IMMEDIATE = auto()
+
+    def is_stop_cmd(self):
+        """Verify if the command is the stop one."""
+        return self in (ControlCommand.STOP, ControlCommand.STOP_IMMEDIATE)
+
+
+class APIWorker(object):
+    """Worker that makes non-blocking HTTP requests to the Report Portal."""
+
+    def __init__(self, cmd_queue, data_queue):
+        """Initialize instance attributes.
+
+        :param cmd_queue:  Queue for the control commands
+        :param data_queue: Queue for the RP requests to process
+        """
+        self._cmd_queue = cmd_queue
+        self._data_queue = data_queue
+        self._thread = None
+        self.name = self.__class__.__name__
+
+    def _command_get(self):
+        """Get control command from the control queue."""
+        try:
+            cmd = self._cmd_queue.get_nowait()
+            logger.debug('[%s] Received {%s} command', self.name, cmd)
+            return cmd
+        except Empty:
+            return None
+
+    def _command_process(self, cmd):
+        """Process control command sent to the worker.
+
+        :param cmd: ControlCommand to be processed
+        """
+        if not cmd:
+            return  # No command received
+
+        logger.debug('[%s] Processing {%s} command', self.name, cmd)
+        if cmd == ControlCommand.REPORT_STATUS:
+            logger.debug('[%s] Current status for tasks is: {%s} unfinished',
+                         self.name, self._data_queue.unfinished_tasks)
+
+        if cmd == ControlCommand.STOP:
+            request = self._request_get()
+            while request is not None:
+                self._request_process(request)
+                request = self._request_get()
+
+        if cmd and cmd.is_stop_cmd():
```

Comment:
`if cmd` case is covered in line 70 and may be removed
In general, how about calling `self._stop()` at the end of line 78: `if cmd == ControlCommand.STOP:`?